File: Java/ReverseVowels.java
package Java;
import java.util.Set;

public class ReverseVowels {
  private static String reverseVowels(String word) {
    char[] letters = word.toCharArray();
    int left = 0;
    int right = letters.length - 1;
    Set<Character> vowels = Set.of('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U');

    while (left < right) {
      if (vowels.contains(letters[left]) && vowels.contains(letters[right])) {
        char temp = letters[left];
        letters[left] = letters[right];
        letters[right] = temp;

        left++;
        right--;
      } else {
        if (!vowels.contains(letters[left])) left++;
        if (!vowels.contains(letters[right])) right--;
      }
    }
    return new String(letters);
  }

  public static void main(String[] args) {
    System.out.println(reverseVowels("amendoim"));
    System.out.println(reverseVowels("IceCreAm"));
    System.out.println(reverseVowels("leetcode"));
  }
}

File: TS/reverseVowels.ts
function reverseVowels(word: string): string {
  let letters = word.split("");
  const vowels = "aeiouAEIOU";
  let left = 0;
  let right = word.length - 1;

  while (left < right) {
    if (vowels.includes(letters[right]) && vowels.includes(letters[left])) {
      [letters[right], letters[left]] = [letters[left], letters[right]];
      left++;
      right--;
    } else {
      if (!vowels.includes(letters[left])) left++;
      if (!vowels.includes(letters[right])) right--;
    }
  }
  return letters.join("");
}

console.log(reverseVowels("amendoim"));
console.log(reverseVowels("IceCreAm"));
console.log(reverseVowels("leetcode"));

/**
 * Versão mais limpa tem um while com um if else dentro
 * Usa two pointer. Se ambos são vogal, troca usando atribuição de array
 * Se não forem, testa cada ponteiro, se não for, move o ponteiro
 */